File: python/agentseal/shield.py
# agentseal/shield.py
"""
Shield - continuous filesystem monitoring for AI agent security.

Watches skill directories, MCP config files, and agent config dirs.
When a file changes, triggers an incremental scan and optionally sends
desktop notifications.

Usage:
    agentseal shield          # foreground, Ctrl+C to stop
    agentseal shield --quiet  # suppress terminal output, notifications only
"""

import signal
import sys
import time
from collections import defaultdict
from pathlib import Path
from threading import Timer
from typing import Callable, Optional

from agentseal.baselines import BaselineStore
from agentseal.guard_models import GuardVerdict
from agentseal.machine_discovery import _get_well_known_configs, _home, _SKILL_DIRS, _SKILL_FILES
from agentseal.mcp_checker import MCPConfigChecker
from agentseal.notify import Notifier
from agentseal.skill_scanner import SkillScanner
from agentseal.toxic_flows import analyze_toxic_flows

try:
    from watchdog.events import FileSystemEvent, FileSystemEventHandler
    from watchdog.observers import Observer

    _WATCHDOG_AVAILABLE = True
except ImportError:
    _WATCHDOG_AVAILABLE = False


# Colors for terminal output
_R = "\033[91m"
_Y = "\033[93m"
_G = "\033[92m"
_D = "\033[90m"
_B = "\033[1m"
_C = "\033[96m"
_RST = "\033[0m"


# Callback type: (event_type, path, result_summary) -> None
ShieldCallback = Callable[[str, str, str], None]


def check_watchdog_available() -> None:
    """Raise ImportError with install instructions if watchdog is missing."""
    if not _WATCHDOG_AVAILABLE:
        raise ImportError(
            "agentseal shield requires the 'watchdog' package.\n"
            "Install with: pip install agentseal[shield]"
        )


class _DebouncedHandler(FileSystemEventHandler):
    """Filesystem event handler with per-path debouncing.

    Accumulates events and fires the callback only after a quiet period
    (no new events for the same path). This prevents redundant scans when
    editors save multiple times or git operations touch many files.
    """

    def __init__(
        self,
        on_change: Callable[[Path], None],
        debounce_seconds: float = 2.0,
    ):
        super().__init__()
        self._on_change = on_change
        self._debounce = debounce_seconds
        self._timers: dict[str, Timer] = {}

    def on_any_event(self, event: "FileSystemEvent") -> None:
        # Skip directory events - we only care about file changes
        if event.is_directory:
            return
        # Skip temporary/swap files from editors
        src = event.src_path
        if src.endswith(("~", ".swp", ".swx", ".tmp", ".DS_Store")):
            return

        path_key = src
        # Cancel any pending timer for this path
        existing = self._timers.pop(path_key, None)
        if existing is not None:
            existing.cancel()

        # Schedule a new scan after the debounce period
        t = Timer(self._debounce, self._fire, args=(path_key,))
        t.daemon = True
        t.start()
        self._timers[path_key] = t

    def _fire(self, path_key: str) -> None:
        self._timers.pop(path_key, None)
        self._on_change(Path(path_key))

    def cancel_all(self) -> None:
        """Cancel all pending timers."""
        for t in self._timers.values():
            t.cancel()
        self._timers.clear()


def _collect_watch_paths() -> tuple[list[Path], list[Path]]:
    """Collect all paths that shield should monitor.

    Returns:
        watch_dirs: Directories to watch recursively (skill dirs)
        watch_files: Individual files to watch (MCP configs, single skill files)
    """
    import platform as _platform

    home = _home()
    system = _platform.system()
    configs = _get_well_known_configs()

    watch_dirs: list[Path] = []
    watch_files: list[Path] = []
    seen: set[str] = set()

    def _add_dir(p: Path) -> None:
        resolved = str(p.resolve())
        if resolved not in seen and p.is_dir() and not p.is_symlink():
            seen.add(resolved)
            watch_dirs.append(p)

    def _add_file(p: Path) -> None:
        resolved = str(p.resolve())
        if resolved not in seen and p.exists():
            seen.add(resolved)
            watch_files.append(p)

    # MCP config files from all known agents
    for cfg in configs:
        path = cfg["paths"].get(system) or cfg["paths"].get("all")
        if path is None:
            continue
        path = Path(path).expanduser()
        # Watch the parent directory (so we catch file creation too)
        if path.parent.is_dir():
            _add_dir(path.parent)

    # Skill directories
    for skill_dir_rel in _SKILL_DIRS:
        skill_dir = home / skill_dir_rel
        _add_dir(skill_dir)

    # Single skill files - watch their parent dirs
    for skill_file_rel in _SKILL_FILES:
        skill_file = home / skill_file_rel
        if skill_file.parent.is_dir():
            _add_dir(skill_file.parent)

    # Current working directory skill files
    try:
        cwd = Path.cwd()
        for name in [".cursorrules", "CLAUDE.md", ".github"]:
            candidate = cwd / name
            if candidate.exists():
                if candidate.is_dir():
                    _add_dir(candidate)
                else:
                    _add_file(candidate)
    except OSError:
        pass

    return watch_dirs, watch_files


def _classify_path(path: Path) -> str:
    """Classify a changed file as 'skill', 'mcp_config', or 'unknown'."""
    name = path.name.lower()
    suffix = path.suffix.lower()

    # MCP config files
    mcp_names = {
        "claude_desktop_config.json",
        "mcp.json",
        "mcp_config.json",
        "cline_mcp_settings.json",
    }
    if name in mcp_names:
        return "mcp_config"

    # Agent settings that may contain MCP config
    if name in ("settings.json", "config.json") and any(
        part in str(path).lower()
        for part in (".claude", ".cursor", ".gemini", ".codex", ".kiro", ".opencode",
                     ".continue", ".aider", ".roo", ".amp", "windsurf", "zed")
    ):
        return "mcp_config"

    # Skill files
    if suffix in (".md", ".txt", ".yaml", ".yml"):
        return "skill"
    if name in (".cursorrules",):
        return "skill"

    return "unknown"


class Shield:
    """Continuous filesystem monitor for AI agent security.

    Watches skill directories and MCP config files for changes.
    When a change is detected, runs an incremental scan on just the
    changed file and optionally sends a desktop notification.
    """

    def __init__(
        self,
        semantic: bool = True,
        notify: bool = True,
        debounce_seconds: float = 2.0,
        on_event: Optional[ShieldCallback] = None,
    ):
        check_watchdog_available()
        self._semantic = semantic
        self._debounce = debounce_seconds
        self._on_event = on_event or (lambda *a: None)
        self._notifier = Notifier(enabled=notify, min_interval=30.0)
        self._scanner = SkillScanner(semantic=semantic)
        self._mcp_checker = MCPConfigChecker()
        self._baseline_store = BaselineStore()
        self._observer: Optional["Observer"] = None
        self._running = False
        self._scan_count = 0
        self._threat_count = 0

    @property
    def scan_count(self) -> int:
        return self._scan_count

    @property
    def threat_count(self) -> int:
        return self._threat_count

    def _handle_change(self, path: Path) -> None:
        """Handle a single file change event."""
        if not path.is_file():
            return

        file_type = _classify_path(path)
        self._scan_count += 1

        if file_type == "skill":
            self._scan_skill(path)
        elif file_type == "mcp_config":
            self._scan_mcp_config(path)
        else:
            # Unknown file type in a watched directory - try skill scan
            if path.suffix.lower() in (".md", ".txt", ".yaml", ".yml"):
                self._scan_skill(path)

    def _scan_skill(self, path: Path) -> None:
        """Run incremental skill scan on a single file."""
        try:
            result = self._scanner.scan_file(path)
        except Exception:
            self._on_event("error", str(path), "Failed to scan file")
            return

        if result.verdict == GuardVerdict.DANGER:
            self._threat_count += 1
            detail = result.top_finding.title if result.top_finding else "Threat detected"
            self._on_event("threat", str(path), f"DANGER - {detail}")
            self._notifier.notify_threat(
                item_name=result.name,
                item_type="Skill",
                severity=result.top_finding.severity if result.top_finding else "high",
                detail=detail,
            )
        elif result.verdict == GuardVerdict.WARNING:
            detail = result.top_finding.title if result.top_finding else "Warning"
            self._on_event("warning", str(path), f"WARNING - {detail}")
        else:
            self._on_event("clean", str(path), "CLEAN")

    def _scan_mcp_config(self, path: Path) -> None:
        """Re-check MCP servers when a config file changes."""
        import json
        from agentseal.machine_discovery import _strip_json_comments

        try:
            raw = path.read_text(encoding="utf-8")
            data = json.loads(_strip_json_comments(raw))
        except (json.JSONDecodeError, OSError):
            self._on_event("error", str(path), "Failed to parse config")
            return

        # Try common MCP keys
        servers: dict = {}
        for key in ("mcpServers", "servers", "context_servers"):
            if key in data and isinstance(data[key], dict):
                servers = data[key]
                break

        if not servers:
            self._on_event("clean", str(path), "No MCP servers in config")
            return

        has_threat = False
        server_dicts = []
        for srv_name, srv_cfg in servers.items():
            if not isinstance(srv_cfg, dict):
                continue
            server_dict = {"name": srv_name, "source_file": str(path), **srv_cfg}
            server_dicts.append(server_dict)

            # MCP config check
            result = self._mcp_checker.check(server_dict)
            if result.verdict == GuardVerdict.DANGER:
                has_threat = True
                self._threat_count += 1
                detail = result.top_finding.title if result.top_finding else "Threat detected"
                self._on_event("threat", str(path), f"MCP '{srv_name}': DANGER - {detail}")
                self._notifier.notify_threat(
                    item_name=srv_name,
                    item_type="MCP Server",
                    severity=result.top_finding.severity if result.top_finding else "high",
                    detail=detail,
                )
            elif result.verdict == GuardVerdict.WARNING:
                detail = result.top_finding.title if result.top_finding else "Warning"
                self._on_event("warning", str(path), f"MCP '{srv_name}': WARNING - {detail}")

            # Baseline check (rug pull detection)
            change = self._baseline_store.check_server(server_dict)
            if change and change.change_type in ("config_changed", "binary_changed"):
                self._threat_count += 1
                self._on_event("warning", str(path), f"BASELINE: {change.detail}")
                self._notifier.notify_threat(
                    item_name=srv_name,
                    item_type="MCP Baseline",
                    severity="high",
                    detail=change.detail,
                )

        # Toxic flow analysis across all servers in this config
        if len(server_dicts) >= 2:
            flows = analyze_toxic_flows(server_dicts)
            for flow in flows:
                self._on_event("warning", str(path), f"TOXIC FLOW: {flow.title}")

        if not has_threat:
            self._on_event("clean", str(path), f"MCP config OK ({len(servers)} servers)")

    def start(self) -> tuple[int, int]:
        """Start watching. Returns (dirs_watched, files_watched).

        This sets up the observer but does NOT block. Call run_forever()
        or use start()/stop() for manual control.
        """
        check_watchdog_available()

        watch_dirs, watch_files = _collect_watch_paths()
        handler = _DebouncedHandler(
            on_change=self._handle_change,
            debounce_seconds=self._debounce,
        )
        self._handler = handler

        self._observer = Observer()

        watched_count = 0
        for d in watch_dirs:
            try:
                self._observer.schedule(handler, str(d), recursive=True)
                watched_count += 1
            except OSError:
                pass  # Permission denied or path disappeared

        # For individual files, watch their parent directory
        file_parents: set[str] = set()
        for f in watch_files:
            parent = str(f.parent)
            if parent not in file_parents:
                file_parents.add(parent)
                try:
                    self._observer.schedule(handler, parent, recursive=False)
                    watched_count += 1
                except OSError:
                    pass

        self._observer.start()
        self._running = True
        return watched_count, len(watch_files)

    def stop(self) -> None:
        """Stop the filesystem observer."""
        self._running = False
        if hasattr(self, "_handler"):
            self._handler.cancel_all()
        if self._observer is not None:
            self._observer.stop()
            self._observer.join(timeout=5)
            self._observer = None

    def run_forever(self) -> None:
        """Block until interrupted (Ctrl+C or SIGTERM)."""
        stop_requested = False

        def _signal_handler(signum, frame):
            nonlocal stop_requested
            stop_requested = True

        signal.signal(signal.SIGINT, _signal_handler)
        signal.signal(signal.SIGTERM, _signal_handler)

        try:
            while not stop_requested and self._running:
                time.sleep(0.5)
        finally:
            self.stop()

File: js/src/shield.ts
/**
 * Shield — continuous filesystem monitoring for AI agent security.
 *
 * Watches skill directories, MCP config files, and agent config dirs.
 * When a file changes, triggers an incremental scan and optionally sends
 * desktop notifications.
 *
 * Uses Node.js built-in fs.watch (recursive on macOS/Windows).
 *
 * Port of Python agentseal/shield.py.
 */

import { existsSync, readFileSync, statSync, watch, type FSWatcher } from "node:fs";
import { homedir } from "node:os";
import { basename, dirname, extname, join, resolve } from "node:path";

import { BaselineStore } from "./baselines.js";
import { Blocklist } from "./blocklist.js";
import { GuardVerdict, type SkillFinding } from "./guard-models.js";
import {
  getWellKnownConfigs,
  stripJsonComments,
  PROJECT_SKILL_DIRS,
  PROJECT_SKILL_FILES,
} from "./machine-discovery.js";
import { MCPConfigChecker } from "./mcp-checker.js";
import { Notifier } from "./notify.js";
import { SkillScanner } from "./skill-scanner.js";
import { analyzeToxicFlows } from "./toxic-flows.js";
import { scanSkillFile, computeVerdict } from "./guard.js";

// ═══════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════

/** Callback: (eventType, path, resultSummary) */
export type ShieldCallback = (eventType: string, path: string, summary: string) => void;

export interface ShieldOptions {
  /** Enable semantic analysis. Default: false */
  semantic?: boolean;
  /** Enable desktop notifications. Default: true */
  notify?: boolean;
  /** Debounce seconds for filesystem events. Default: 2.0 */
  debounceSeconds?: number;
  /** Callback for shield events. */
  onEvent?: ShieldCallback;
}

// ═══════════════════════════════════════════════════════════════════════
// DEBOUNCED HANDLER
// ═══════════════════════════════════════════════════════════════════════

/**
 * Per-path debouncing for filesystem events.
 * Accumulates events and fires only after a quiet period.
 */
export class DebouncedHandler {
  private _onChange: (filePath: string) => void;
  private _debounceMs: number;
  private _timers = new Map<string, ReturnType<typeof setTimeout>>();

  constructor(onChange: (filePath: string) => void, debounceMs = 2000) {
    this._onChange = onChange;
    this._debounceMs = debounceMs;
  }

  /** Handle a filesystem event. Skips directories and temp files. */
  handleEvent(filePath: string, isDirectory = false): void {
    if (isDirectory) return;

    // Skip temp/swap files from editors
    if (
      filePath.endsWith("~") ||
      filePath.endsWith(".swp") ||
      filePath.endsWith(".swx") ||
      filePath.endsWith(".tmp") ||
      filePath.endsWith(".DS_Store")
    ) {
      return;
    }

    // Cancel existing timer for this path
    const existing = this._timers.get(filePath);
    if (existing !== undefined) {
      clearTimeout(existing);
    }

    // Schedule new scan after debounce
    const timer = setTimeout(() => {
      this._timers.delete(filePath);
      this._onChange(filePath);
    }, this._debounceMs);

    this._timers.set(filePath, timer);
  }

  /** Cancel all pending timers. */
  cancelAll(): void {
    for (const timer of this._timers.values()) {
      clearTimeout(timer);
    }
    this._timers.clear();
  }

  /** Number of pending timers (for testing). */
  get pendingCount(): number {
    return this._timers.size;
  }
}

// ═══════════════════════════════════════════════════════════════════════
// PATH CLASSIFICATION
// ═══════════════════════════════════════════════════════════════════════

const MCP_CONFIG_NAMES = new Set([
  "claude_desktop_config.json",
  "mcp.json",
  "mcp_config.json",
  "cline_mcp_settings.json",
]);

const AGENT_PATH_MARKERS = [
  ".claude", ".cursor", ".gemini", ".codex", ".kiro", ".opencode",
  ".continue", ".aider", ".roo", ".amp", "windsurf", "zed",
];

/** Classify a changed file as 'skill', 'mcp_config', or 'unknown'. */
export function classifyPath(filePath: string): string {
  const name = basename(filePath).toLowerCase();
  const ext = extname(filePath).toLowerCase();

  // MCP config files
  if (MCP_CONFIG_NAMES.has(name)) return "mcp_config";

  // Agent settings that may contain MCP config
  if ((name === "settings.json" || name === "config.json")) {
    const lower = filePath.toLowerCase();
    if (AGENT_PATH_MARKERS.some((marker) => lower.includes(marker))) {
      return "mcp_config";
    }
  }

  // Skill files
  if ([".md", ".txt", ".yaml", ".yml"].includes(ext)) return "skill";
  if (name === ".cursorrules") return "skill";

  return "unknown";
}

// ═══════════════════════════════════════════════════════════════════════
// WATCH PATH COLLECTION
// ═══════════════════════════════════════════════════════════════════════

function isDir(p: string): boolean {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function fileExists(p: string): boolean {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

/** Collect all paths that shield should monitor. */
export function collectWatchPaths(homeOverride?: string): { dirs: string[]; files: string[] } {
  const home = homeOverride ?? homedir();
  const plat =
    process.platform === "darwin"
      ? "Darwin"
      : process.platform === "win32"
        ? "Windows"
        : "Linux";

  const configs = getWellKnownConfigs();

  const dirs: string[] = [];
  const files: string[] = [];
  const seen = new Set<string>();

  const addDir = (p: string) => {
    const resolved = resolve(p);
    if (!seen.has(resolved) && isDir(p)) {
      seen.add(resolved);
      dirs.push(p);
    }
  };

  const addFile = (p: string) => {
    const resolved = resolve(p);
    if (!seen.has(resolved) && fileExists(p)) {
      seen.add(resolved);
      files.push(p);
    }
  };

  // MCP config files from all known agents — watch parent directories
  for (const cfg of configs) {
    const paths = cfg.paths as Record<string, string | undefined>;
    let cfgPath = paths[plat] ?? paths.all;
    if (!cfgPath) continue;
    cfgPath = cfgPath.replace(/^~/, home);
    const parent = dirname(cfgPath);
    if (isDir(parent)) addDir(parent);
  }

  // Skill directories
  for (const skillDirRel of PROJECT_SKILL_DIRS) {
    const skillDir = join(home, skillDirRel);
    addDir(skillDir);
  }

  // Single skill files — watch parent dirs
  for (const skillFileRel of PROJECT_SKILL_FILES) {
    const skillFile = join(home, skillFileRel);
    const parent = dirname(skillFile);
    if (isDir(parent)) addDir(parent);
  }

  // CWD skill files
  try {
    const cwd = process.cwd();
    for (const name of [".cursorrules", "CLAUDE.md", ".github"]) {
      const candidate = join(cwd, name);
      if (isDir(candidate)) addDir(candidate);
      else if (fileExists(candidate)) addFile(candidate);
    }
  } catch {
    // ignore
  }

  return { dirs, files };
}

// ═══════════════════════════════════════════════════════════════════════
// SHIELD CLASS
// ═══════════════════════════════════════════════════════════════════════

export class Shield {
  private _onEvent: ShieldCallback;
  private _notifier: Notifier;
  private _scanner: SkillScanner;
  private _mcpChecker: MCPConfigChecker;
  private _blocklist: Blocklist;
  private _baselineStore: BaselineStore;
  private _debounceMs: number;
  private _watchers: FSWatcher[] = [];
  private _handler: DebouncedHandler | null = null;
  private _running = false;
  private _scanCount = 0;
  private _threatCount = 0;

  constructor(options: ShieldOptions = {}) {
    this._onEvent = options.onEvent ?? (() => {});
    this._notifier = new Notifier(options.notify ?? true);
    this._scanner = new SkillScanner();
    this._mcpChecker = new MCPConfigChecker();
    this._blocklist = new Blocklist();
    this._baselineStore = new BaselineStore();
    this._debounceMs = (options.debounceSeconds ?? 2.0) * 1000;
  }

  get scanCount(): number {
    return this._scanCount;
  }

  get threatCount(): number {
    return this._threatCount;
  }

  get running(): boolean {
    return this._running;
  }

  /** Handle a single file change event. */
  handleChange(filePath: string): void {
    if (!fileExists(filePath)) return;

    const fileType = classifyPath(filePath);
    this._scanCount++;

    if (fileType === "skill") {
      this._scanSkill(filePath);
    } else if (fileType === "mcp_config") {
      this._scanMcpConfig(filePath);
    } else {
      // Unknown file type — try skill scan for text-like files
      const ext = extname(filePath).toLowerCase();
      if ([".md", ".txt", ".yaml", ".yml"].includes(ext)) {
        this._scanSkill(filePath);
      }
    }
  }

  private _scanSkill(filePath: string): void {
    try {
      const result = scanSkillFile(filePath, this._scanner, this._blocklist);

      if (result.verdict === GuardVerdict.DANGER) {
        this._threatCount++;
        const detail = result.findings[0]?.title ?? "Threat detected";
        this._onEvent("threat", filePath, `DANGER - ${detail}`);
        this._notifier.notifyThreat(
          result.name,
          "Skill",
          result.findings[0]?.severity ?? "high",
          detail,
        );
      } else if (result.verdict === GuardVerdict.WARNING) {
        const detail = result.findings[0]?.title ?? "Warning";
        this._onEvent("warning", filePath, `WARNING - ${detail}`);
      } else {
        this._onEvent("clean", filePath, "CLEAN");
      }
    } catch {
      this._onEvent("error", filePath, "Failed to scan file");
    }
  }

  private _scanMcpConfig(filePath: string): void {
    let data: Record<string, any>;
    try {
      const raw = readFileSync(filePath, "utf-8");
      data = JSON.parse(stripJsonComments(raw));
    } catch {
      this._onEvent("error", filePath, "Failed to parse config");
      return;
    }

    // Try common MCP keys
    let servers: Record<string, any> = {};
    for (const key of ["mcpServers", "servers", "context_servers"]) {
      if (key in data && typeof data[key] === "object" && data[key] !== null) {
        servers = data[key];
        break;
      }
    }

    if (Object.keys(servers).length === 0) {
      this._onEvent("clean", filePath, "No MCP servers in config");
      return;
    }

    let hasThreat = false;
    const serverDicts: Array<Record<string, any>> = [];

    for (const [srvName, srvCfg] of Object.entries(servers)) {
      if (typeof srvCfg !== "object" || srvCfg === null) continue;
      const serverDict = { name: srvName, source_file: filePath, ...srvCfg };
      serverDicts.push(serverDict);

      // MCP config check
      const result = this._mcpChecker.check(serverDict);
      if (result.verdict === GuardVerdict.DANGER) {
        hasThreat = true;
        this._threatCount++;
        const detail = result.findings[0]?.title ?? "Threat detected";
        this._onEvent("threat", filePath, `MCP '${srvName}': DANGER - ${detail}`);
        this._notifier.notifyThreat(
          srvName,
          "MCP Server",
          result.findings[0]?.severity ?? "high",
          detail,
        );
      } else if (result.verdict === GuardVerdict.WARNING) {
        const detail = result.findings[0]?.title ?? "Warning";
        this._onEvent("warning", filePath, `MCP '${srvName}': WARNING - ${detail}`);
      }

      // Baseline check (rug pull detection)
      const change = this._baselineStore.checkServer(serverDict);
      if (change && (change.change_type === "config_changed" || change.change_type === "binary_changed")) {
        this._threatCount++;
        this._onEvent("warning", filePath, `BASELINE: ${change.detail}`);
        this._notifier.notifyThreat(srvName, "MCP Baseline", "high", change.detail);
      }
    }

    // Toxic flow analysis across all servers
    if (serverDicts.length >= 2) {
      const flows = analyzeToxicFlows(serverDicts);
      for (const flow of flows) {
        this._onEvent("warning", filePath, `TOXIC FLOW: ${flow.title}`);
      }
    }

    if (!hasThreat) {
      this._onEvent("clean", filePath, `MCP config OK (${Object.keys(servers).length} servers)`);
    }
  }

  /**
   * Start watching. Returns { dirsWatched, filesWatched }.
   *
   * Uses Node.js fs.watch with recursive option (macOS/Windows).
   * Does NOT block — call stop() to clean up.
   */
  start(homeOverride?: string): { dirsWatched: number; filesWatched: number } {
    const { dirs, files } = collectWatchPaths(homeOverride);

    this._handler = new DebouncedHandler(
      (fp) => this.handleChange(fp),
      this._debounceMs,
    );

    let watchedCount = 0;

    for (const d of dirs) {
      try {
        const watcher = watch(d, { recursive: true }, (_eventType, filename) => {
          if (filename) {
            this._handler?.handleEvent(join(d, filename));
          }
        });
        this._watchers.push(watcher);
        watchedCount++;
      } catch {
        // Permission denied or path disappeared
      }
    }

    // For individual files, watch their parent directory (non-recursive)
    const fileParents = new Set<string>();
    for (const f of files) {
      const parent = dirname(f);
      if (!fileParents.has(parent)) {
        fileParents.add(parent);
        try {
          const watcher = watch(parent, { recursive: false }, (_eventType, filename) => {
            if (filename) {
              this._handler?.handleEvent(join(parent, filename));
            }
          });
          this._watchers.push(watcher);
          watchedCount++;
        } catch {
          // Permission denied
        }
      }
    }

    this._running = true;
    return { dirsWatched: watchedCount, filesWatched: files.length };
  }

  /** Stop the filesystem watchers. */
  stop(): void {
    this._running = false;
    if (this._handler) {
      this._handler.cancelAll();
      this._handler = null;
    }
    for (const w of this._watchers) {
      try {
        w.close();
      } catch {
        // ignore
      }
    }
    this._watchers = [];
  }
}
